### PR TITLE
Add TileLink interconnect generator

### PR DESCRIPTION
**Pull request by zhemao**
Add a tilelink interconnect generator. This is more-or-less copy-pasted from the AXI generator in junctions, then modified to use TileLink instead.

### CODE REVIEW

File: src/main/scala/interconnect.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+      f"Region $name start address 0x$start%x before previous region end")
+
+    realAddrMap(i) = (start, size)
+    lastEnd = start + size
+  }
+
+  val routeSel = (addr: UInt) => {
+    Vec(realAddrMap.map { case (start, size) =>
+      addr >= UInt(start) && addr < UInt(start + size)
+    }).toBits
```
Comment (aswaterman) [inline src/main/scala/interconnect.scala]: Don't create Vecs unless you're actually using them for dynamic indexing.  In this case, you could do

    val foo = realAddrMap.map { ... }
    Cat(foo.reverse:_*)

File: src/main/scala/interconnect.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      Module(new ClientUncachedTileLinkIORouter(nOuter, routeSel)) }
+    val arbiters = List.fill(nOuter) {
+      Module(new ClientUncachedTileLinkIOArbiter(nInner)) }
+
+    for (i <- 0 until nInner) {
+      routers(i).io.in <> io.in(i)
+    }
+
+    for (i <- 0 until nOuter) {
+      arbiters(i).io.in <> Vec(routers.map(r => r.io.out(i)))
```
Comment (aswaterman) [inline src/main/scala/interconnect.scala]: Don't create Vecs unless you're actually using them for dynamic indexing.  In this case, you should just be able <> without using the Vec at all, right?

**Comment by aswaterman** [comment]
modulo line notes, I think this is fine

**Comment by zhemao** [comment]
Okay, I've made some updates.

**Comment by hcook** [comment]
Merged.

**Pull request closed without merging**